### ISSUE TITLE
Use mtime to avoid checking content hashes for repository up-to-dateness checks

### ISSUE THREAD

**Issue by lberki** [bazelbuild/bazel#21044]
### Description of the feature request:

The way Bazel verifies up-to-dateness of source files is that if the mtime is unchanged, it believes it, and if not, it does a secondary check by checksumming source files to avoid re-running actions after a simple `touch`.

Repository rules don't work like this: they simply checksum every file that the repository depends on, which makes up-to-dateness much slower than they would otherwise be. 

It would be nice to make repository rules do the same thing as actions.

### What underlying problem are you trying to solve with this feature?

Make the repository up-to-dateness check more efficient.

**Comment by fmeum** [comment]
Just want to note that Bazel doesn't check just mtime, but also ctime and other attributes, before computing a digest: https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/vfs/DigestUtils.java;l=42;bpv=1;bpt=1

**Comment by lberki** [comment]
You're indeed right -- I used "mtime" as a shorthand for "whatever can be gleaned about a file short of looking at its contents".

**Comment by SalmaSamy** [comment]
@lberki, @fmeum I guess storing only ctime should be enough? as it already tracks changes to the file content and the extra stuff

**Comment by lberki** [comment]
@SalmaSamy AFAIU the ctime always changes if mtime does, so using that instead of the mtime would cause suboptimal caching in the worst case, which is perfectly fine (if somewhat inconsistent with action inputs, but that's just a minor gripe)

**Comment by fmeum** [comment]
There is the (probably very theoretical) concern of having the filesystem backing a path change. By chance, the ctime could remain the same but the file contents could differ. Bazel users now stat properties such as file size and inside for the digest cache key, but I could see that being overkill for repo rules.

**Comment by lberki** [comment]
Good point! TBH I think the best approach here is to mimic what `FileContentsProxy` does, which is a tuple of (mtime, ctime, inode number)., It's only very slightly more complicated than relying on ctime alone, and protects against a few odd scenarios like the one you describe above.

**Comment by meteorcloudy** [comment]
@bazel-io fork 7.1.0

**Comment by meisterT** [comment]
Note that timestamps on HFS+ have a granularity of one second.

**Comment by lberki** [comment]
FWIW, that's what `TimestampGranularityMonitor` is for: it makes Bazel wait long enough at the end of the build so that Bazel can be sure that if something modifies an input after it has returned, its timestamp changes.

**Comment by Wyverald** [comment]
Lowering the priority of this one due to the uncertainty of its correctness (and we're out of time for 7.1.0 anyway).
